src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+                Object oldValue;
+                if (v == null) {
+                    oldValue = nil;
+                } else {
+                    oldValue = v;
+                }
+
+                Object newValue = callBlock(block, oldValue);
+
+                if (newValue == nil) {
+                    return null;
+                } else {
+                    return newValue;
+                }
+            });
+
+            if (returnValue == null) {
+                return nil;
+            }
+            return returnValue;
+        }
+    }
+
+    @CoreMethod(names = "merge_pair", required = 2, needsBlock = true)
+    public abstract static class MergePairNode extends YieldingCoreMethodNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object mergePair(RubyConcurrentHashMap self, Object key, Object value, RubyProc block) {
+            Object returnValue = self.concurrentHash.merge(new RubyConcurrentHashMap.Key(key), value, (k, v) -> {
+                final Object oldValue = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+                Object newValue = callBlock(block, oldValue);
+
+                if (newValue == nil) {
+                    return null;
+                } else {
+                    return newValue;
+                }
+            });
+
+            if (returnValue == null) {
+                return nil;
+            }
+            return returnValue;
+        }
+    }
+
+    @CoreMethod(names = "replace_pair", required = 3)
+    public abstract static class ReplacePairNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected boolean replacePair(RubyConcurrentHashMap self, Object key, Object oldValue, Object newValue) {
+            return self.concurrentHash.replace(new RubyConcurrentHashMap.Key(key), oldValue, newValue);
+        }
+    }
+
+    @CoreMethod(names = "replace_if_exists", required = 2)
+    public abstract static class ReplaceIfExistsNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object replaceIfExists(RubyConcurrentHashMap self, Object key, Object newValue) {
+            final Object oldValue = self.concurrentHash.replace(new RubyConcurrentHashMap.Key(key), newValue);
+            if (oldValue == null) {
+                return nil;
+            } else {
+                return oldValue;
+            }
+        }
+    }
+
+    @CoreMethod(names = "get_and_set", required = 2)
+    public abstract static class GetAndSetNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object getAndSet(RubyConcurrentHashMap self, Object key, Object value) {
+            final Object oldValue = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+            self.concurrentHash.put(new RubyConcurrentHashMap.Key(key), value);
```

Comment:
This is not atomic, it should only use `put()` and the return value of it.
`CHM#put()` returns the previous value.

Reply:
Fixed in [`86889e5` (#2339)](https://github.com/oracle/truffleruby/pull/2339/commits/86889e533ba18fe72e55b753cb8c1300787b90f0)